src/compiler/checker.ts

```diff
@@ -25203,6 +25203,48 @@ namespace ts {
             return literalType;
         }
 
+        /**
+         * Replaces all references to structurally equivalent object and array literal types in the given list
+         * with references to a single one of those types. The process is applied recursively to properties of
+         * object literals and elements of array literals.
+         */
+        function deduplicateObjectOrArrayLiteralTypes(types: Type[]) {
+            if (!some(types, isObjectOrArrayLiteralType)) {
+                return types;
+            }
+            const typeMap = new Map<string, Type>();
+            return sameMap(types, getInternedType);
+
+            function getInternedType(type: Type): Type {
+                if (type.flags & TypeFlags.Union) {
+                    const newTypes = sameMap((<UnionType>type).types, getInternedType);
+                    return newTypes !== (<UnionType>type).types ? getUnionType(newTypes) : type;
+                }
+                const key = getLiteralTypeKey(type);
+                return key ? typeMap.get(key) || (typeMap.set(key, type), type) : type;
+            }
+
+            function getLiteralTypeKey(type: Type): string | undefined {
+                const objectFlags = getObjectFlags(type);
+                if (objectFlags & ObjectFlags.ObjectLiteral) {
+                    const props = getPropertiesOfObjectType(type);
+                    const propTypes = map(props, p => getInternedType(getTypeOfSymbol(p)));

```

Comment:
This is going to more eagerly pull on symbols with deferred types; I'm unsure how often `SymbolFlags.Deferred` arises in places this interning occurs, but perhaps we should respect the deferral flag and, like, just use the symbol ID as the key in those scenarios?

Reply:
We won't pull on anything unless it is marked `ObjectFlags.ObjectLiteral` or `ObjectFlags.ArrayLiteral` and those are never deferred.